### DISCUSSION TITLE
Show Device State Vs Show Device Statistics Mismatch

### DISCUSSION THREAD

**Discussion by LucasTraumen** [IchthysMaranatha/asterisk-chan-quectel#6] [Q&A] [answered]
First I wanted to say BIG thanks for your thorough readme I def have a better understanding of the Asterisk/Chan_quectel interaction. 

There seems to be an issue with tracking and maintaining the state of my Quectel Object. For the enqueued commands on the "Show Device State" and the "Show Device Statistics" be the same? This may be a "feature" inherited from Chan_dongle as well, just wanted to double-check.

Background, I'm using the Quectel EC25-AF and I am able to send calls/receive calls also.
I'm Currently Running Asterisk 18.8 as well on Ubuntu (Updated for any passers-by)

SBC-Modem*CLI> quectel show device state quectel0
-------------- Status -------------
  Device                  : quectel0
  State                   : Free
  Audio                   : /dev/ttyUSB1
  Data                    : /dev/ttyUSB3
  Voice                   : Yes
  SMS                     : Yes
  Manufacturer            : Quectel
  Model                   : EC25
  Firmware                : 
  IMEI                    : XXXXXXXXXXXXXXX
  IMSI                    : XXXXXXXXXXXXXXX
  GSM Registration Status : Registered, home network
  RSSI                    : 24, -65 dBm
  Mode                    : 
  Submode                 :
  Provider Name           : XXXX
  Location area code      : "XXXX"
  Cell ID                 : "xxxxxx
  Subscriber Number       : +xxxxxxxxxxx
  SMS Service Center      : +xxxxxxxxxxx
  Use UCS-2 encoding      : Yes
  Tasks in queue          : 0
  Commands in queue       : 0
  Call Waiting            : Disabled
  Current device state    : start
  Desired device state    : start
  When change state       : now
  Calls/Channels          : 0
    Active                : 0
    Held                  : 0
    Dialing               : 0
    Alerting              : 0
    Incoming              : 0
    Waiting               : 0
    Releasing             : 0
    Initializing          : 0

SBC-Modem*CLI> quectel show device statistics  quectel0
-------------- Statistics -------------
  Device                      : quectel0
  Queue tasks                 : 2
  Queue commands              : 21
  Responses                   : 34
  Bytes of read responses     : 390
  Bytes of written commands   : 217
  Bytes of read audio         : 0
  Bytes of written audio      : 0
  Readed frames               : 0
  Readed short frames         : 0
  Wrote frames                : 0
  Wrote short frames          : 0
  Wrote silence frames        : 0
  Write buffer overflow bytes : 0
  Write buffer overflow count : 0
  Incoming calls              : 0
  Waiting calls               : 0
  Handled input calls         : 0
  Fails to PBX run            : 0
  Attempts to outgoing calls  : 0
  Answered outgoing calls     : 0
  Answered incoming calls     : 0
  Seconds of outgoing calls   : 0
  Seconds of incoming calls   : 0
  ACD for incoming calls      : -1
  ACD for outgoing calls      : -1
  ASR for incoming calls      : -1
  ASR for outgoing calls      : -1

**Comment by IchthysMaranatha** [comment, marked as answer]
I haven't done any changes in the stats code after I forked this. There is an issue where I'm forcing registration due to an error in parsing of Quectel CREG with quotation marks. 

I've corrected that and will be uploading the changes soon. But I don't think that will affect the device stats. Looking at it again, it may very well be that state is supposed to show current state which shows no commands/tasks in queue currently, while stats show all that have been placed before in queue (and may have been already executed).

**Reply by LucasTraumen** [reply]
In chan_quectel.h line 167 cell_id[8]; -> cell_id[10]; This makes the Cell ID from : "XXXXXX to : "XXXXXXXX". 8 truncates CREG string

**Reply by IchthysMaranatha** [reply]
Thanks for pointing that out but that alone won't fix location and cell id issues. chan-dongle code assumes a max of 4 parameters for CREG and if response contains 1 or 3 of them it assumes this is URC and if 2 or 4 it assumes it's a response to command. Whereas quectel will always have 4 or more parameters. You'll see in the device state after placing or receiving a call, the location code will contain truncated cell id and and cell id will show the network type parameter. I'll update with a fix.

**Reply by IchthysMaranatha** [reply]
Have updated code. Values may show a trailing double quote
